scripts/src/java/org/oppia/android/scripts/coverage/CoverageRunner.kt

```diff
@@ -41,4 +53,176 @@ class CoverageRunner(
   ): List<String>? {
     return bazelClient.runCoverageForTestTarget(bazelTestTarget)
   }
+
+  private fun parseCoverageData(
+    coverageData: List<String>,
+    bazelTestTarget: String
+  ): CoverageReport {
+    var filePath = ""
+    var linesFound = 0
+    var linesHit = 0
+    val coveredLines = mutableListOf<CoveredLine>()
```

Comment:
Mutability here isn't strictly bad, but it could be avoided by taking multiple passes on the data file lines (and specifically extracting out lines with the prefixes that are needed for each thing). It's definitely not as performant, and that could matter for large files.

For values where we need to guarantee the value is set, I highly suggest using lateinit var and not specifying a default, or specifying a null default and checking for nullability at the end (to ensure all required values are set).

Reply:
Another option could be to generically parse the structure since it seems well-defined as properties, and then construct the sub-elements of the proto by pulling from the generic parsed structure. In this case, I think you could parse all the lines into a ``Map<String, List<List<String>>>`` data structure using something like:

```kotlin
val coverageDataProps = coverageData.groupBy { line ->
  val (key, value) = line.split(":", limit = 2)
  return@groupBy key
}.mapValues { (_, line) ->
  line.substringAfter(":").split(",")
}
```

Then it can become pretty straightforward:

```kotlin
val functionsFound = coverageDataProps.getValue("FNF").single().single().toInt()
val coveredLines = coverageDataProps.getValue("DA").map { (lineNumStr, hitCountStr) ->
  CoveredLine.newBuilder().apply {
    this.lineNumber = lineNumStr.toInt()
    this.coverage = if (hitCountStr.toInt() > 0) Coverage.FULL else Coverage.NONE
  }.build()
}
```

As a couple of examples. This shows cases where:
- Lines must be present in the data file (FNF, DA).
- A case where exactly one entry should be present (FNF).
- A case where an entry has exactly one property (FNF).
- A case where a line can occur multiple times in the file (DA).
- Using destructuring to extract properties from a multi-property line (DA).

This could be another approach to eliminate the mutability _and_ nesting, introduce stronger assertion checks (e.g. with ``getValue()`` and ``single()``), and do this without a performance hit (since the map structure should be well optimized for this type of repeated retrieval and iteration pattern unlike the raw lines).

Reply:
:astonished: thanks @BenHenning, for just giving the entire parsing logic. Updated the source with the same!